### PR TITLE
Add comments for the "Constructing the form data set" algorithm

### LINKED ISSUE #7852: Document the specification steps in get_form_dataset
`get_form_dataset` in `HTMLFormElement.rs` is implementing [this algorithm](https://html.spec.whatwg.org/multipage/forms.html#constructing-the-form-data-set) from the HTML specification, but we lack comments labeling each being implemented. We should label this method like we do [other ones](http://mxr.mozilla.org/servo/source/components/script/dom/document.rs#1333) for clarity.

Code: components/script/dom/htmlformelement.rs

### PR DESCRIPTION
**Pull request by askobara**
Fixes #7852

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8955)

### CODE REVIEW

**Comment by Wafflespeanut** [comment]
Nice, this should be helpful. Some nits...

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/htmlinputelement.rs, line 445 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8955#-K5QzbsOrDlPNPrji3gl:-K5QzbsOrDlPNPrji3gm:-1235803009) ([raw file](https://github.com/servo/servo/blob/3151de203c665b824d271fba1cdeed239e853245/components/script/dom/htmlinputelement.rs#L445)):*
Nit: I think it'd be clean if we have the `if` inside the match arm, and move the comment to the top of the arm.

---

*[components/script/dom/htmlinputelement.rs, line 448 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8955#-K5Qzq2UUO1ezh0l56Qd:-K5Qzq2VFF7s4VMksLkS:1132024844) ([raw file](https://github.com/servo/servo/blob/3151de203c665b824d271fba1cdeed239e853245/components/script/dom/htmlinputelement.rs#L448)):*
Nit: ditto (since there's only one `if`), and we can forget the braces :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8955#-:-K5QzW-kcQB6ZJPArFH_:2018495514)*

**Comment by nox** [comment]
```
/home/travis/build/servo/servo/components/script/dom/htmlinputelement.rs:440:9: 448:10 error: non-exhaustive patterns: `Atom { .. }` not covered [E0004]
/home/travis/build/servo/servo/components/script/dom/htmlinputelement.rs:440         match ty {
/home/travis/build/servo/servo/components/script/dom/htmlinputelement.rs:441             // Step 3.1: it's a button but it is not submitter.
/home/travis/build/servo/servo/components/script/dom/htmlinputelement.rs:442             atom!("submit") | atom!("button") | atom!("reset") if !is_submitter => return None,
/home/travis/build/servo/servo/components/script/dom/htmlinputelement.rs:443             // Step 3.1: it's the "Checkbox" or "Radio Button" and whose checkedness is false.
/home/travis/build/servo/servo/components/script/dom/htmlinputelement.rs:444             atom!("radio") | atom!("checkbox") if !self.Checked() || name.is_empty() => return None,
/home/travis/build/servo/servo/components/script/dom/htmlinputelement.rs:445             atom!("image") | atom!("file") => return None, // Unimplemented
                                                                             ...
/home/travis/build/servo/servo/components/script/dom/htmlinputelement.rs:440:9: 448:10 help: run `rustc --explain E0004` to see a detailed explanation
```

File: components/script/dom/htmlinputelement.rs [hunk 1]
```diff
@@ -434,21 +438,17 @@ impl HTMLInputElement {
...
+            // Step 3.1: it's the "Checkbox" or "Radio Button" and whose checkedness is false.
+            atom!("radio") | atom!("checkbox") if !self.Checked() || name.is_empty() => return None,
             atom!("image") | atom!("file") => return None, // Unimplemented
-            _ => {
-                if name.is_empty() {
-                    return None;
-                }
-            }
+            // Step 3.1: it's not the "Image Button" and doesn't have a name attribute.
+            _ if name.is_empty() => return None
```
Comment (KiChjang) [inline components/script/dom/htmlinputelement.rs]: This part here is what's generating the compilation error. You removed the `_` match case and added a pattern guard here.

This has different semantics than the code you deleted. Previously it meant "match against all other values of `ty`, and if name is empty, return None". Right now it means "match against all other values of `ty` that has their name empty, and return None". See the difference?

**Comment by Wafflespeanut** [comment]
Yep, sorry I should've been more specific. I meant something like,

``` rust
match foo {
    foobar => (),
    _ => if thing {
        return None;
    }
}
```

**Comment by askobara** [comment]
yeah, sorry guys, I'm hurried
thanks @KiChjang for the detailed answer and @Wafflespeanut for the example of code, now I hope everything is fine

**Comment by Manishearth** [comment]
Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[components/script/dom/htmlinputelement.rs, line 445 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8955#-K5QzbsOrDlPNPrji3gl:-K5U9LSqeF_K_cEfFxAW:1957926790) ([raw file](https://github.com/servo/servo/blob/3151de203c665b824d271fba1cdeed239e853245/components/script/dom/htmlinputelement.rs#L445)):*
If you move the if inside the match arm, the match arm is unconditionally executed. We want it to be able to try the next arms if is_submitter is true

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8955)*

**Comment by Wafflespeanut** [comment]
@Manishearth You forgot to mention that the current changes are fine :)

**Pull request merged**